### DISCUSSION TITLE
better sound , more realistic

### DISCUSSION THREAD

**Discussion by StronSon** [libsidplayfp/libsidplayfp#44] [Ideas]
RESID_INLINE
    int output(const WaveformGenerator* ringModulator)// const
    {
		//wgc changed 01/02.2022
		//return static_cast<int>(waveformGenerator->output(ringModulator) * envelopeGenerator->output()) *master_volume;//WGCCHANGE

		//wgc changed 01/02.2022
		fix++;
		if (fix > 2.) fix = 1;//=0;
		float out = ((waveformGenerator->output(ringModulator) * envelopeGenerator->output()) *(fix*0.5))*master_volume;//WGCCHANGE

		// not needed but just incase 
		if (out > 3.40282e+038) { out = 3.40282e+038; }else	if (out < 1.17549e-038) out = 1.17549e-038;
		
		return static_cast<int>(out);
    }

give it a try and let me know what you think , im comparing against 4 different real sids , this seems to produce a much better sounding output, a lot closer to real hardware , sounds a lot less synthetic and a lot more analog.

**Comment by StronSon** [comment]
https://soundcloud.com/the-original-755589985/sanxion-1?in=the-original-755589985/sets/c64-sanxion-and-others

**Comment by drfiemost** [comment]
There's no appreciable difference for me, but my hearing is pretty bad so I'd better check a spectrogram.
Anyway, what's the rationale behind this? It's just a quick volume wobble.

**Reply by Toddler-Boy** [reply]
Did you ever have time to investigate this? It seems weird as the volume alternates between 50% and 100% for every second sample, and the clamping right after should be completely unnecessary. Maybe it should be written as an assertion, to test if it's needed?

**Reply by drfiemost** [reply]
As [mywave82](https://github.com/mywave82) says below it would produce a tone at half of the sampling frequency (so 500KHz) which would be then cut out by the filter.

**Reply by Toddler-Boy** [reply]
It doesn’t generate a tone. It modulates the volume of the ringmodulator output. So it should only be audible in tunes that use the ring modulator feature. I will modify my local copy and compare with Sanxion and a couple others that make heavy use of the ring mod.

**Reply by drfiemost** [reply]
It's not that, `ringModulator` is just a pointer to the ring modulating voice, so `(waveformGenerator->output(ringModulator) * envelopeGenerator->output())` is the output of a voice modulated by the envelope generator.

**Reply by Toddler-Boy** [reply]
So it’s modulating the volume of the voice then. I guess you could call that a high frequency ring modulation in itself then. I will add this “fix” and see if I can notice an appreciable difference.

**Reply by Toddler-Boy** [reply]
This does nothing except drop the volume. When compensated for, the resulting wav-files cancel each other out.

**Comment by mywave82** [comment]
Will not this code generate different output depending on the sampling rate. A 44100 sampling rate soundcard will generate a 22050 Hz tone while a 96000 sampling rate soundcard generate 48000Hz tone ?

**Reply by drfiemost** [reply]
No, the chip is clocked at around 1MHz, the output is resampled at a later stage.

**Comment by Toddler-Boy** [comment]
@StronSon

>less synthetic and a lot more analog

Except that the real SID chip produces a digital output signal that optionally gets filtered by a semi-analog filter. libsidplayfp emulates that near perfectly by rendering the voices at 1 MHz and then downsampling to the target sample-rate.

Your code just lowers the overall volume a bit, but otherwise has no effect on the frequency spectrum.
